src/storage.rs

```diff
@@ -75,10 +92,24 @@ pub trait Storage {
     /// the slice of entries returned will always have length at least 1 if entries are
     /// found in the range.
     ///
+    /// If the entries are fetched asynchorously, it would return LogTemporarilyUnavailable,
+    /// and application needs to call `on_entries_fetched(context)` to trigger re-fetch of the entries
+    /// after the storage finishes fetching the entries.   
+    ///
+    /// Constraint: only [low, high] covers unapplied entries are permitted to fetch entries
```

Comment:
It may be a burden for the user to determine `[low, high] covers unapplied entries`, and there may also have use cases that required fetching unapplied logs synchronously in the future.

How about providing a helper function like `GetEntriesContext::can_async` instead? /cc @BusyJay

Reply:
It's discussed before that changing the same function from asynchronous to synchronous depends on the parameters can be weird. Unapplied logs are usually stored in memory, so it's unlikely need to be fetched asynchronously. Note synchronous version is always compatible with the asynchronous version, so even it's supported in the future, no breaking change is introduced.

Reply:
> It's discussed before that changing the same function from asynchronous to synchronous depends on the parameters can be weird. 

Using sync or async never depends on the parameter but the user, namely how the user implements `Storage::entries`. The user can always ignore the `GetEntriesContext` parameter and fetch logs in sync way, if the user wants the performance gain by fetch logs asynchronously, `GetEntriesContext::can_async` can tell whether `raft-rs` support doing so.  /cc @tonyxuqqi 

> Unapplied logs are usually stored in memory, so it's unlikely need to be fetched asynchronously.

Again, it is very up to the user's implementation of the `Storage` trait.

> Note synchronous version is always compatible with the asynchronous version, so even it's supported in the future, no breaking change is introduced.

What I mean is the usage inside `raft-rs`.  For example, `self.raft_log.entries(l, h, ctx).unwrap()` where `[l. h]` only covers unapplied entries so the user may think it is okay to fetch logs asynchronously, and thus get an unexpected panic. Whether fetching log asynchronously is allowed is up to the usage inside `raft-rs` not other conditions.

Reply:
> ...where [l. h] only covers unapplied entries...

However such usage doesn't exist. Raft query unapplied logs only when sending logs to followers.

Reply:
Even the correct constraint can still meet the abovementioned problem.

Reply:
ping

Reply:
`can_async` is quite like the previous implementation which has the issue that changing the same function from asynchronous to synchronous depends on the parameters, though I'm still not sure if it's really a problem. And the constraint is not a good way either in my point of view. Both are acceptable for me now. 😑

Reply:
> It could easily be broken by future use cases or future use cases may have to subject to this unreasonable constraint.

My suggestion is to remove the constraint in the end.

Reply:
Encounter a problem that some other places TiKV utilize `entries()` as well. They need to pass a context to make sure not to use async way. So `can_async` is needed. I think we waste too much time on it and it makes everything complicated to avoid the so-called style problem.

Reply:
I checked the code, and it's indeed hard to solve. Perhaps we do need to expose `can_async` and context definition, at least constructor. I'm now OK with the changes.

/cc @tonyxuqqi what's your suggestion?

Reply:
IMO, there are two ways to implement this feature, either integrated into `Storage::entries` or a new interface like `Storage::entries_async`.

With `Storage::entries`, as there are cases that allow using the async way and cases that **must** use the sync way, we must pass some information to let `Storage::entries` know what the caller expected.  And I think helper function like `can_async` is a more clear, straightforward and robust way to do so. (actually, the abovementioned constraint also rely on the parameter, namely the `[low, high]` range)

With `Storage::entries_async`, although make the `Storage` trait look more redundant, both `entries` and `entries_async` keep in simple and much clear than any parameter. With a default implement like:
```
fn entries_async(...) {
    self.entries(...)
}
```
there are only very small changes to the codebase, with no breaking change in the interface and backward compatibility.

Both ways are okay to me.

Reply:
After all the callers support async way, the `entries_async` is verbose. So I choose to provide a `can_async`.